### DISCUSSION TITLE
Script file not be loaded

### DISCUSSION THREAD

**Discussion by woohung** [netbox-community/netbox#12813] [Help Wanted!] [answered]
Hello, folks!
Before updating to 3.5.2, I used a custom script to poll devices with the ping command and change the status in netbox, depending on the ping response.

In 3.5.2, as I understand it, the approach to working with scripts has changed and everything now works through the GUI.
When adding a script, I get the following status:
Script file at: /opt/netbox/netbox/scripts/check_device.py could not be loaded.
The problem is exactly in my script, since from examples on adding a new Site, the script is successfully added.
I`m new in netbox scripts, help me figure out how to debug stuff like this?
I will be grateful for help. Thank you.

Code:
```python
# Modules
from icmplib import ping


# Netbox artifcats
from extras.scripts import Script, ObjectVar, MultiObjectVar
from dcim.models import Site, Device


# Classes
class DeviceChecker(Script):
    class Meta:
        name = "Validate state of network devices"
        description = "Sample Script to set the status of the device based on its availability"
        field_order = ("site", "devices")

    site = ObjectVar(
        model=Site
    )

    devices = MultiObjectVar(
        model=Device,
        query_params={
            "site_id": "$site"
        }
    )

    def run(self, data, commit) -> None:
        print_in_nb = []
        for device in data["devices"]:
            result = ping(str(device.primary_ip.address.ip), count=3, interval=0.2, privileged=False)
       
            if result.is_alive:
                device.status = "active"

            else:
                device.status = "offline"

            print_in_nb.append(f"{device.name} is set to {device.status}")
            device.save()

        return "\n".join(print_in_nb)
```

**Comment by candlerb** [comment, marked as answer]
> Script file at: /opt/netbox/netbox/scripts/check_device.py could not be loaded.

Yes, debugging scripts is currently a pain if they have an error which prevents them from being loaded in the first place.

I gave a workaround at #12766. Basically you take a copy of your script, insert the following lines at the top of it:

```
#!/opt/netbox/venv/bin/python
import django, os, sys
sys.path.append('/opt/netbox/netbox')
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'netbox.settings')
django.setup()
```

Then run it using `/opt/netbox/venv/bin/python myscript.py` (or you can do `chmod +x myscript.py` then `./myscript.py`). Errors which prevent the script being loaded should then be shown.

If I test your script this way, I get

```
ModuleNotFoundError: No module named 'icmplib'
```

because icmplib isn't in my virtualenv. I installed the required library temporarily using `/opt/netbox/venv/bin/pip install icmplib` and then it was happy.

The long-term solution to this is for you to create `/opt/netbox/local_requirements.txt` containing just the line `icmplib`, and then whenever you run `./upgrade.sh` it will include this package when rebuilding the python virtualenv (i.e. every time you upgrade).

**Reply by woohung** [reply]
yes, I immediately thought about icmplib and even thought that I installed it.
The command `/opt/netbox/venv/bin/pip install icmplib` solved the problem, in my case I also added `sudo`.
Please help me to understand, why command only worked like this `sudo /opt/netbox/venv/bin/pip install icmplib`?
What is the difference from `sudo pip install icmplib`?
both commands were executed from a virtual environment.

**Reply by candlerb** [reply]
> Please help me to understand, why command only worked like this `sudo /opt/netbox/venv/bin/pip install icmplib`?
> What is the difference from `sudo pip install icmplib`?

When Netbox runs, it's running inside a virtualenv and can only see and use packages installed inside that virtualenv.

`/opt/netbox/venv/bin/pip install icmplib` installs packages inside the virtualenv. You can think of it as a shortcut to

```
source /opt/netbox/venv/activate
pip install icmplib
deactivate
```

`pip install icmplib` installs libraries into a python global package directory.  These packages are visible to a python interpreter which is running *outside* of a virtualenv.

**Reply by woohung** [reply]
Yes, I understood this part, but in my case it did not help. The first thing I did was just:
``` 
source /opt/netbox/venv/bin/activate 
sudo pip install icmplib 
``` 
`sudo pip list` showed the icmplib module, but the script didn't work. It only worked after:
`sudo /opt/netbox/venv/bin/pip install icmplib`

**Reply by woohung** [reply]
By the way, output `sudo /opt/netbox/venv/bin/pip list` and `sudo pip list` is different.
And yes, i'm tryed `sudo pip list` after source `source /opt/netbox/venv/bin/activate`

**Reply by candlerb** [reply]
Your "sudo" is getting rid of the environment variables which activate the virtualenv.

It will work if you start a root shell, and activate the virtualenv *inside* that root shell:

```
sudo -s
source /opt/netbox/venv/activate
pip install icmplib
exit
```

Otherwise, you might get it to work with `sudo -E` which preserves the environment, and/or `-H` which sets the HOME directory variable. The exact set of flags would need to be tested.  I always use the root shell to be sure it's correct.

You could test:

```
source /opt/netbox/venv/bin/activate 
sudo -E pip list
sudo -EH pip list
```

(EDIT: I tried these, and on my system, neither shows the packages in the Netbox virtualenv. Please just use one of the other approaches which works)

(EDIT 2: I found a [working recipe](https://askubuntu.com/questions/234758/how-to-use-a-python-virtualenv-with-sudo#answer-1342154): `sudo -E env PATH="$PATH" pip list`)

**Reply by woohung** [reply]
oh, yes, `sudo -s` works. I'm understood, thx! I'll go read more about `sudo`.

(EDIT: yes, `sudo -E env PATH="$PATH" pip list` works too)
